lwcomponent/http.go

```diff
@@ -19,58 +19,89 @@
 package lwcomponent
 
 import (
-	"io"
-	"net/http"
+	"fmt"
+	"log"
 	"os"
 	"time"
+
+	"github.com/go-resty/resty/v2"
 )
 
 const (
-	defaultTimeout  = 30 * time.Second
-	defaultMaxRetry = 2
+	DefaultMaxRetry = 3
+	defaultTimeout  = 5 * time.Minute
 )
 
-func DownloadFile(filepath string, url string, timeout time.Duration) (err error) {
+// Default timeout is 5 minutes
+// Retry 3 times (4 requests total)
+// Resty default RetryWaitTime is 100ms
+// Exponential backoff to a maximum of RetryWaitTime of 2s
+func DownloadFile(filepath string, url string, size int64, timeout time.Duration) (err error) {
+	client := resty.New()
+
 	if timeout == 0 {
 		timeout = defaultTimeout
 	}
 
-	client := &http.Client{Timeout: timeout}
+	client.SetTimeout(timeout)
+	client.SetRetryCount(DefaultMaxRetry)
 
-	file, err := os.Create(filepath)
+	client.OnError(func(req *resty.Request, err error) {
+		if v, ok := err.(*resty.ResponseError); ok {
+			fmt.Println(v.Response.Body())
+		}
+	})
+
+	_, err = os.Create(filepath)
 	if err != nil {
-		return err
+		return
 	}
-	defer file.Close()
 
-	err = downloadFile(client, url, file)
+	done := make(chan int64)
 
-	for retry := 0; retry < defaultMaxRetry && os.IsTimeout(err); retry++ {
-		_, err = file.Seek(0, io.SeekStart)
-		if err != nil {
-			return err
-		}
+	go downloadProgress(done, filepath, size)
 
-		err = downloadFile(client, url, file)
-	}
+	response, err := client.R().SetOutput(filepath).Get(url)
+
+	done <- response.Size()
 
 	return
 }
 
-// client.Get returns on receiving HTTP headers and we stream the HTTP data to the output file.
-// A timeout will interrupt io.Copy.
-func downloadFile(client *http.Client, url string, file *os.File) (err error) {
-	var (
-		resp *http.Response
-	)
-
-	resp, err = client.Get(url)
+func downloadProgress(done chan int64, filepath string, totalSize int64) {
+	file, err := os.Open(filepath)
 	if err != nil {
-		return err
+		log.Fatal(err)
 	}
-	defer resp.Body.Close()
 
-	_, err = io.Copy(file, resp.Body)
+	defer file.Close()
 
-	return
+	var stop bool = false
+
+	for !stop {
+		select {
+		case <-done:
+			stop = true
+		default:
+			info, err := file.Stat()
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			size := info.Size()
+			if size == 0 {
+				size = 1
+			}
+
+			if totalSize == 0 {
+				fmt.Printf("Downloaded: %.0fkb\n", float64(size)/(1<<10))
```

Comment:
The `lwcomponent` should not interact with STDOUT or STDERR

Reply:
How should we indicate progress?

Reply:
I am now using a virtual terminal - is this the approach you had in mind?